File: automations/seo-optimizer/supabase/functions/seo-optimizer-analyst/categories/low-ctr.ts
// Category C: Low CTR — top-10 con CTR muy bajo (problema de title/meta).

import { ctrBenchmarkForPosition, trafficPotentialFromLowCtr } from "../../_shared/scoring.ts";
import type { ClientAnalysisContext, OpportunityCandidate } from "../types.ts";

const MIN_POSITION = 1;
const MAX_POSITION = 10;
const MIN_IMPRESSIONS = 1000;
const CTR_GAP_THRESHOLD = 0.5;

export function detect(ctx: ClientAnalysisContext): OpportunityCandidate[] {
  const byUrl = new Map<string, typeof ctx.gscRows>();
  for (const r of ctx.gscRows) {
    if (r.position < MIN_POSITION || r.position > MAX_POSITION) continue;
    if (r.impressions < MIN_IMPRESSIONS) continue;
    const expected = ctrBenchmarkForPosition(r.position);
    if (expected <= 0) continue;
    if (r.ctr < expected * CTR_GAP_THRESHOLD) {
      if (!byUrl.has(r.url)) byUrl.set(r.url, []);
      byUrl.get(r.url)!.push(r);
    }
  }

  const out: OpportunityCandidate[] = [];
  for (const [url, rows] of byUrl) {
    let totalPotential = 0;
    const underperforming = rows.map(r => {
      const expected = ctrBenchmarkForPosition(r.position);
      const potential = trafficPotentialFromLowCtr({
        impressionsMonthly: r.impressions, actualCtr: r.ctr, position: r.position,
      });
      totalPotential += potential;
      return {
        query: r.query,
        position: Number(r.position.toFixed(2)),
        impressions: r.impressions,
        ctr_actual: Number(r.ctr.toFixed(4)),
        ctr_expected: Number(expected.toFixed(4)),
        ctr_gap: Number((expected - r.ctr).toFixed(4)),
        potential_clicks_at_benchmark: Number(potential.toFixed(1)),
      };
    }).sort((a, b) => b.potential_clicks_at_benchmark - a.potential_clicks_at_benchmark);

    if (underperforming.length === 0) continue;
    const primaryQuery = underperforming[0].query;
    const ci = ctx.contentItemByUrl.get(url);

    out.push({
      category: "low_ctr",
      articleUrl: url,
      articleTitle: ci?.title ?? "",
      articleLanguage: ci?.language ?? null,
      contentItemId: ci?.contentItemId ?? null,
      evidence: {
        underperforming_queries: underperforming.slice(0, 8),
        total_potential_clicks_monthly: Number(totalPotential.toFixed(1)),
      },
      recommendationSummary:
        `CTR muy bajo para la posición. ${underperforming.length} queries pierden ` +
        `~${totalPotential.toFixed(0)} clicks/mes. Reescribir title tag y meta description.`,
      recommendationDetails: {
        action: "rewrite_title_and_meta",
        current_title: ci?.title ?? null,
        current_meta: ci?.metaDescription ?? null,
        focus_queries: underperforming.slice(0, 5).map(q => q.query),
        tactics: [
          "Incluir la query principal al inicio del title",
          "Usar números, year actual, o paréntesis para destacar",
          "Meta description en primera persona del lector, con beneficio claro y CTA",
          "Evitar truncamiento (50-60 chars title, 150-160 meta)",
        ],
      },
      trafficPotential: totalPotential,
      confidence: "high",
      effortLevel: "low",
      dedupeSignature: `${url}:meta:${primaryQuery}`,
    });
  }
  return out;
}

File: automations/seo-optimizer/02-agents/analyst/categories/low_ctr.py
"""Category C: Low CTR — top-10 position but CTR < 50% of benchmark.

This is a TITLE / META problem, not content. Cheapest fix in SEO.

Detect rule:
  - position in [1, 10]
  - impressions >= 1000/month
  - ctr_actual < ctr_expected × 0.5

Score = clicks_left_on_table × 2.5 (super high: low effort + clear ROI).
Confidence: high.  Effort: low.
"""
from __future__ import annotations

from collections import defaultdict

from _shared.scoring import ctr_benchmark_for_position, traffic_potential_from_low_ctr
from analyst.types import ClientAnalysisContext, OpportunityCandidate


MIN_POSITION = 1.0
MAX_POSITION = 10.0
MIN_IMPRESSIONS = 1000
CTR_GAP_THRESHOLD = 0.5  # actual < benchmark × this triggers detection


def detect(ctx: ClientAnalysisContext) -> list[OpportunityCandidate]:
    by_url: dict[str, list] = defaultdict(list)
    for r in ctx.gsc_rows:
        if not (MIN_POSITION <= r.position <= MAX_POSITION):
            continue
        if r.impressions < MIN_IMPRESSIONS:
            continue
        expected = ctr_benchmark_for_position(r.position)
        if expected <= 0:
            continue
        if r.ctr < expected * CTR_GAP_THRESHOLD:
            by_url[r.url].append(r)

    out: list[OpportunityCandidate] = []
    for url, rows in by_url.items():
        # Aggregate across queries: total clicks_left_on_table
        underperforming = []
        total_potential = 0.0
        for r in rows:
            expected = ctr_benchmark_for_position(r.position)
            potential = traffic_potential_from_low_ctr(
                impressions_monthly=r.impressions, actual_ctr=r.ctr, position=r.position,
            )
            total_potential += potential
            underperforming.append({
                "query": r.query,
                "position": round(r.position, 2),
                "impressions": r.impressions,
                "ctr_actual": round(r.ctr, 4),
                "ctr_expected": round(expected, 4),
                "ctr_gap": round(expected - r.ctr, 4),
                "potential_clicks_at_benchmark": round(potential, 1),
            })

        underperforming.sort(key=lambda q: -q["potential_clicks_at_benchmark"])
        primary_query = underperforming[0]["query"] if underperforming else ""

        ci = ctx.content_item_by_url.get(url)
        out.append(OpportunityCandidate(
            category="low_ctr",
            article_url=url,
            article_title=ci.title if ci else "",
            article_language=ci.language if ci else None,
            content_item_id=ci.content_item_id if ci else None,
            evidence={
                "underperforming_queries": underperforming[:8],
                "total_potential_clicks_monthly": round(total_potential, 1),
            },
            recommendation_summary=(
                f"CTR muy bajo para la posición. {len(underperforming)} queries pierden "
                f"~{total_potential:.0f} clicks/mes. Reescribir title tag y meta description."
            ),
            recommendation_details={
                "action": "rewrite_title_and_meta",
                "current_title": ci.title if ci else None,
                "current_meta": ci.meta_description if ci else None,
                "focus_queries": [q["query"] for q in underperforming[:5]],
                "tactics": [
                    "Incluir la query principal al inicio del title",
                    "Usar números, year actual, o paréntesis para destacar",
                    "Meta description en primera persona del lector, con beneficio claro y CTA",
                    "Evitar truncamiento (50-60 chars title, 150-160 meta)",
                ],
            },
            traffic_potential=float(total_potential),
            confidence="high",
            effort_level="low",
            dedupe_signature=f"{url}:meta:{primary_query}",
        ))

    return out
